File: client/result-tree-node.ts
import { RGQLValue } from '../rgraphql.pb.js'
import { ResultTreeHandler } from './result-tree-handler.js'

export class ResultTreeNode {
  // children contains any children of this node
  public children: ResultTreeNode[] = []

  constructor(
    // value is the value of this node.
    public value: RGQLValue,
  ) {}

  // callHandler recursively calls handlers with the contents of the tree.
  // cb is called with each *rtNode and handler combo
  public callHandler(
    handler: ResultTreeHandler,
    cb: (rtn: ResultTreeNode, rth: ResultTreeHandler) => void,
  ) {
    if (!handler) {
      return
    }
    for (const child of this.children) {
      const nextHandler = handler(child.value)
      if (nextHandler) {
        cb(child, nextHandler)
        child.callHandler(nextHandler, cb)
      }
    }
  }
}

File: client/result_tree_node.go
package client

import (
	proto "github.com/rgraphql/rgraphql"
)

// rtNode is a node in the result tree
type rtNode struct {
	value    *proto.RGQLValue
	children []*rtNode
}

// newRtNode builds a new result tree node.
func newRtNode(val *proto.RGQLValue) *rtNode {
	return &rtNode{value: val}
}

// callHandler recursively calls handlers with the contents of the tree.
// cb is called with each *rtNode and handler combo
func (n *rtNode) callHandler(handler ResultTreeHandler, cb func(*rtNode, ResultTreeHandler)) {
	for _, child := range n.children {
		nextHandler := handler.HandleResultValue(child.value)
		if nextHandler != nil {
			cb(child, nextHandler)
			child.callHandler(nextHandler, cb)
		}
	}
}

// func (n *rtNode) findIndex()
